vendor/github.com/ethereum/go-ethereum/p2p/discv5/net.go

```diff
@@ -641,7 +641,18 @@ loop:
 				}()
 			} else {
 				refreshDone = make(chan struct{})
-				net.refresh(refreshDone)
+
+				done := make(chan struct{})
+				net.refresh(done)
+				<-done
+
+				// Refresh again only if there are no seeds.
+				// Also, sleep for some time to prevent from
+				// executing too often.
+				go func() {
+					time.Sleep(time.Millisecond * 20)
```

Comment:
so the main change is actually delaying refresh loop by 20ms. i don't understand why it fixes the problem

Reply:
Because `refreshDone` is used in `case <-refreshDone` and this channel is closed immediately when there are no peers. Thus, it may happen that the outer `select` always calls this block instead of e.g. `case req := <-net.topicRegisterReq:` which should be called when a new peer registers a topic.